[NIXL][BUG FIX] Fix a bug for PD with host_buffer after merging 29665

xuechendi opened the pull request:
<!-- markdownlint-disable -->

## Purpose

After https://github.com/vllm-project/vllm/pull/29665 merged, the PR added `remote_request_id=kv_transfer_params["remote_request_id"],` to `add_new_req`

However, in prefill side, there is no `remote_request_id` in kv_transfer_params

=> This issue only occurs when running PD with host_buffer

## Test Plan

```
PREFILL_BLOCK_SIZE=16 DECODE_BLOCK_SIZE=16 bash tests/v1/kv_connector/nixl_integration/run_accuracy_test.sh --kv_buffer_device cpu
```

without the PR, test will fail immediately
With this PR, there will still be an accuracy issue, which fix is proposed at https://github.com/vllm-project/vllm/pull/30419

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

xuechendi committed:
Fix a bug when creating meta at prefill side for save to host

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

markmc reviewed (changes requested):
Thank you

This is only an issue with `--kv_buffer_device cpu` which I guess we don't test in CI

I don't think `remote_request_id` should be particularly special here - a prefill request should be able to just contain `kv_transfer_params = {'do_remote_decode': True}` IMO

xuechendi commented on vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:
```diff
@@ -238,7 +238,7 @@ def add_new_req(
             remote_block_ids=kv_transfer_params["remote_block_ids"],
             remote_engine_id=kv_transfer_params["remote_engine_id"],
-            remote_request_id=kv_transfer_params["remote_request_id"],
+            remote_request_id=kv_transfer_params.get("remote_request_id", ""),
```
Ok, will do. 
`remote_request_id=kv_transfer_params.get("remote_request_id", None)` => Failed pre-commit mypy, so I used `remote_request_id=kv_transfer_params.get("remote_request_id", "")` instead

xuechendi committed:
update add_new_req by setting remote with default value

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi committed:
Use None as default and ignore mypy

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi commented on vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:
```diff
@@ -236,11 +236,11 @@ def add_new_req(
+            remote_engine_id=kv_transfer_params.get("remote_engine_id"),  # type: ignore
+            remote_request_id=kv_transfer_params.get("remote_request_id"),  # type: ignore
+            remote_host=kv_transfer_params.get("remote_host"),  # type: ignore
+            remote_port=kv_transfer_params.get("remote_port"),  # type: ignore
```
@markmc , do you think this works?
I tried updating ReqMeta with optional, then the mypy check exploded. => Since we only want to align with toy_proxy, so I am hoping to use `# type ignore` , in that case, we can still use None and make mypy happy

markmc committed:
[NIXL] Refactor prefill-only ReqMeta into RemoteMeta

On the decode side, in reqs_to_save, we do not expect or need
any of these remote_ fields - they are for the prefill side
only. Make this more clear by putting these fields in their
own dataclass which is only present on requests in reqs_to_recv.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
The mypy errors are a sign that we can do better here - the current code basically disables type checking

```
   def _add_new_req(
         ...
         kv_transfer_params: dict[str, Any],
         _req = ReqMeta(
             ...
             remote_engine_id=kv_transfer_params["remote_engine_id"]
```

Because the value returned by the dict lookup is `Any`, mypy does no type checking - even though it could be None (and we know it definitely is None on the decode side runtime), mypy doesn't complain that we are assigning it to a `str` field. And what's worse, it's a really subtle way of disabling type checking ... it's not at all obvious to the reader, so someone reading this code might be very surprised to find these fields are `None` at runtime

I've sent you xuechendi/vllm-fork#15 which will enable proper type checking. Please do check it over and make sure it makes sense to you :+1:

xuechendi committed:
Merge pull request #15 from markmc/fix_pd_host_buffer

[NIXL] Refactor prefill-only ReqMeta into RemoteMeta

markmc reviewed (approved):
@NickLucche PTAL, since I've done a bunch of refactoring in the latest version

njhill reviewed (approved):
Thanks @xuechendi @markmc! The RemoteMeta refactoring seems much cleaner!

DarkLight1337 committed:
Merge branch 'main' into dev/fix_pd_host_buffer
